src/pymor/reductors/era.py

```diff
@@ -0,0 +1,160 @@
+            mats = [*mats, self.input_projector(l2)] if l2 else mats
+            s1 = ('lp,', 'l') if l1 else ('', 'p')
+            s2 = (',mk', 'k') if l2 else ('', 'm')
+            self.logger.info('Projecting Markov parameters ...')
+            einstr = s1[0] + 'npm' + s2[0] + '->n' + s1[1] + s2[1]
+            h = np.einsum(einstr, *mats, optimize='optimal')
+        else:
+            h = self.data
+
+        self.logger.info(f'Computing SVD of the {"projected " if l1 or l2 else ""}Hankel matrix ...')
+        U, sv, V = spla.svd(to_matrix(NumpyHankelOperator(h)), full_matrices=False)
+
+        return sv, U.T, V
+
+    def output_projector(self, l1):
+        self.logger.info(f'Constructing output projector ({l1} tangential directions) ...')
+        assert isinstance(l1, int) and l1 <= self.data.shape[1]
+        return self._s1_W1()[1][:, :l1]
+
+    def input_projector(self, l2):
+        self.logger.info(f'Constructing input projector ({l2} tangential directions) ...')
+        assert isinstance(l2, int) and l2 <= self.data.shape[2]
+        return self._s2_W2()[1][:, :l2]
+
+    def error_bounds(self, l1=None, l2=None):
+        sv = self._sv_U_V(l1, l2)[0]
+        _, p, m = self.data.shape
+        err = (np.sqrt(np.arange(len(sv)) + (l2 or m) + (l1 or p)) * sv)[1:]
+        err = 2 * err if l1 or l2 else err
+        if l1:
+            s1 = self._s1_W1()[0]
+            err += 4 * np.linalg.norm(s1[l1:])**2
+        if l2:
+            s2 = self._s2_W2()[0]
+            err += 4 * np.linalg.norm(s2[l2:])**2
+        return err
+
+    def reduce(self, r=None, tol=None, l1=None, l2=None):
+        """Construct a minimal realization.
+
+
+
+        Parameters
+        ----------
+        r
+            Order of the reduced model if `tol` is `None`, maximum order if `tol` is specified.
+        tol
+            Tolerance for the error bound if `r` is `None`.
+        l1
+            Number of left (output) directions for tangential projection.
+        l2
+            Number of right (input) directions for tangential projection.
+
+        Returns
+        -------
+        rom
+            Reduced-order discrete-time |LTIModel|.
+        """
+        assert r is not None or tol is not None
+        s, p, m = self.data.shape
+        assert l1 is None or isinstance(l1, int) and l1 <= p
+        assert l2 is None or isinstance(l2, int) and l2 <= m
+        assert r is None or 0 < r <= min((l1 or p), (l2 or m)) * (s // 2 + 1)
+
+        sv, U, V = self._sv_U_V(l1=l1, l2=l2)
+
+        if tol is not None:
+            error_bounds = self.error_bounds(l1=l1, l2=l2)
+            r_tol = np.argmax(error_bounds <= tol) + 1
+            r = r_tol if r is None else min(r, r_tol)
+
+        sv, U, V = sv[:r], U[:r], V[:r]
+
+        self.logger.info(f'Constructing reduced realization of order {r} ...')
+        sqS = np.diag(np.sqrt(sv))
+        Zo = U.T @ sqS
+        A = NumpyMatrixOperator(spla.pinv(Zo[: -(l1 or p)]) @ Zo[(l1 or p):])
```

Comment:
It would be better to use `spla.lstsq`.

Reply:
Thanks, I did not think at all about that!

 Could you tell me why? If I looked it up correctly, `pinv` computes an economy SVD and `lstsq` with the default driver does the same. Is it only about avoiding the additional matrix product afterwards or would the xGELSY driver have some benefits as well?

Reply:
The issue is that `pinv` computes the pseudoinverse unnecessarily, which leads to extra computation (especially when the economy SVD is of low rank) and loss in accuracy. I'm not so familiar with different LAPACK drivers, the default seems ok...